ehsm_kms_service/router.js

```diff
@@ -0,0 +1,161 @@
+const { ehsm_keySpec_t, ehsm_keyorigin_t } = require('./ehsm_kms_params.js')
+const { cryptographic_apis, enroll_apis } = require('./apis')
+const logger = require('./logger')
+const {
+  napi_result,
+  _result,
+  create_app_info,
+  store_cmk,
+} = require('./function')
+
+/**
+ *
+ * @param {string} id (keyid|ukeyid)
+ * @returns query (keyid|ukeyid) parameter
+ *
+ */
+const cmk_db_query = (id) => {
+  return {
+    selector: {
+      _id: `cmk:${id}`,
+    },
+    fields: ['keyBlob', 'creator', 'expireTime'],
+    limit: 1,
+  }
+}
+/**
+ * Verify whether the request parameter appid is equal to creator in the database .
+ * Verify expireTime , the expiration time cannot be less than the current time
+ * @param {string} appid
+ * @param {string} keyid
+ * @param {object} res
+ * @param {object} DB
+ * @returns cmk_base64
+ */
+const find_cmk_by_keyid = async (appid, keyid, res, DB) => {
+  const query = cmk_db_query(keyid)
+  const cmk = await DB.partitionedFind('cmk', query)
+  if (cmk.docs.length == 0) {
+    res.send(_result(400, 'keyid error'))
+    return false
+  }
+  if (appid != cmk.docs[0].creator) {
+    res.send(_result(400, 'appid error'))
+    return false
+  }
+  if (new Date().getTime() > cmk.docs[0].expireTime) {
+    res.send(_result(400, 'keyid expire'))
+    return
+  }
+  return cmk.docs[0].keyBlob
+}
+
+const router = async (p) => {
+  const { req, res, DB } = p
+  const { appid, payload } = req.body
+  const action = req.query.Action
+  switch (action) {
+    case enroll_apis.RA_GET_API_KEY:
+      create_app_info(DB, res)
+      break
+    case cryptographic_apis.CreateKey:
+      try {
+        let { keyspec, origin } = payload
+        /**
+         * keyspec、origin convert to enum type
+         * enum in thie ehsm_kms_params.js file
+         */
+        keyspec = ehsm_keySpec_t[keyspec]
+        origin = ehsm_keyorigin_t[origin]
+        const napi_res = napi_result(action, res, [keyspec, origin])
+        napi_res && store_cmk(napi_res, res, appid, payload, DB)

```

Comment:
In napi_result(), if this error is caught:
**catch (e) {
    return _result(400, 'Parsing error')
}**

will store_cmk() be executed ?

Reply:
no,

Reply:
const _result = (code, msg, data = {}) => {
  return {
    code: code,
    message: msg,
    result: {
      ...data,
    },
  }
}

function napi_result(action, res, params) {
  try {
    const napi_res = ehsm_napi[`NAPI_${action}`](...params)
    if (JSON.parse(napi_res).code != 200) {
      res.send(napi_res)
      return false
    } else {
      return JSON.parse(napi_res)
    }
    // })
  } catch (e) {
    **return _result(400, 'Parsing error')**
  }
}

if any exception is caught, napi_result() returns **{code:400, message:'Parsing error'}** to napi_res.
napi_res is not false, so that **&& store_cmk(napi_res, res, appid, payload, DB)** will be executed.

Is my understanding correct?

Reply:
If the code of the napi result is not equal to 200, the result is directly returned to the user
If the code of the napi  result is equal to 200, the result is returned to the called function